rt.zig

```diff
@@ -0,0 +1,276 @@
+// 3. Create a function with in the RuntimeStep struct that will perform the test. At least this
+//    should use `self.get_msg()` which will get the serial log lines from the OS. Look at
+//    test_init or test_panic for examples.
+// 4. In the create function, add your test mode and test function to the switch.
+// 5. Celebrate if it works lel
+
+/// The enumeration of tests with all the runtime tests.
+pub const TestMode = enum {
+    /// This is for the default test mode. This will just run the OS normally.
+    None,
+
+    /// Run the OS's initialisation runtime tests to ensure the OS is properly set up.
+    Initialisation,
+
+    /// Run the panic runtime test.
+    Panic,
+
+    ///
+    /// Return a string description for the test mode provided.
+    ///
+    /// Argument:
+    ///     IN mode: TestMode - The test mode.
+    ///
+    /// Return: []const u8
+    ///     The string description for the test mode.
+    ///
+    pub fn getDescription(mode: TestMode) []const u8 {
+        return switch (mode) {
+            .None => "Runs the OS normally (Default)",
+            .Initialisation => "Initialisation runtime tests",
+            .Panic => "Panic runtime tests",
+        };
+    }
+};
+
+/// The runtime step for running the runtime tests for the OS.
+pub const RuntimeStep = struct {
+    /// The Step, that is all you need to know
+    step: Step,
+
+    /// The builder pointer, also all you need to know
+    builder: *Builder,
+
+    /// The message queue that stores the log lines
+    msg_queue: Queue,
+
+    /// The qemu process, this is needed for the `read_logs` thread.
+    os_proc: *ChildProcess,
+
+    /// The argv of the qemu process so can create the qemu process
+    argv: [][]const u8,
+
+    /// The test function that will be run for the current runtime test.
+    test_func: TestFn,
+
+    /// The error set for the RuntimeStep
+    const Error = error{
+        /// The error for if a test fails. If the test function returns false, this will be thrown
+        /// at the wnd of the make function as we need to clean up first. This will ensure the
+        /// build fails.
+        TestFailed,
+
+        /// This is used for `self.get_msg()` when the queue is empty after a timeout.
+        QueueEmpty,
+    };
+
+    /// The type of the test function.
+    const TestFn = fn (self: *RuntimeStep) bool;
+
+    /// The time used for getting message from the message queue. This is in milliseconds.
+    const queue_timeout: usize = 5000;
+
+    ///
+    /// This will just print all the serial logs.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *RuntimeStep - Self.
+    ///
+    /// Return: bool
+    ///     This will always return true
+    ///
+    fn print_logs(self: *RuntimeStep) bool {
+        while (true) {
+            const msg = self.get_msg() catch return true;
+            defer self.builder.allocator.free(msg);
+        }
+    }
```

Comment:
This doesn't seem to actually print anything :)

Reply:
oops, I forgot to add the warn :P